Bug 1530714 - Report ASRouter errors in ASRouterAdmin

piatra opened the pull request:
This adds a new section in ASRouterAdmin where you can see all logged errors.

<img width="1305" alt="image" src="https://user-images.githubusercontent.com/810040/54942333-c42ac680-4f26-11e9-9e5a-b9885d453fc6.png">

I've broken down the `_remoteLoader` to report missing messages, response failures or failures to parse json. Also all calls to `Cu.reportError` now go through the same function that captures them to be displayed in the admin panel.

k88hudson commented on lib/ASRouter.jsm:
```diff
@@ -49,6 +49,16 @@ const STARTPAGE_VERSION = "6";
+    this._errors = [{timestamp: new Date(), error: e}, ...this._errors];
+  },
+
+  get errors() {
```
I like the `reportError` strategy, but personally I would return the error through `loadMessagesForProvider` so the state can be centralized in `ASRouter._state`, or maybe even as a property inside provider in `ASRouter._state.providers` itself?

Keeping it here makes cleanup more complicated IMO

k88hudson commented on lib/ASRouter.jsm:
```diff
@@ -104,17 +114,29 @@ const MessageLoaderUtils = {
+        MessageLoaderUtils.reportError(e);
+      }
+      if (
+        response &&
```
This might be a little too strict, personally I usually go with `status >= 200 && status < 400`

piatra commented:
Updated the PR
* We now only report `remote` provider errors
* `errors` are part of the `provider` state
* reverted some of the `Cu.reportError` changes so we store provider errors and everything else is reported to the console
* reporting `provider` id now with the errors
<img width="1045" alt="image" src="https://user-images.githubusercontent.com/810040/55014846-9d7e9580-4fe3-11e9-8b60-89604da3b7f5.png">

k88hudson commented on lib/ASRouter.jsm:
```diff
@@ -104,17 +116,26 @@ const MessageLoaderUtils = {
+      if (
+        response &&
+        (response.status >= 200 && response.status < 400) &&
+        (response.ok || response.status === 302)
```
You don't need `=== 302` if you have `<400`

k88hudson commented on content-src/components/ASRouterAdmin/ASRouterAdmin.jsx:
```diff
@@ -562,6 +562,37 @@ export class ASRouterAdminInner extends React.PureComponent {
+  renderErrorMessage({id, errors}) {
+    const providerId = <td rowSpan={errors.length}>{id}</td>;
+    // .reverse() so that the last error (most recent) is first
+    return errors.reverse().map(({error, timestamp}, cellKey) => (<tr key={cellKey}>
```
can you `.reverse` after `.map` (don't mutate input)
